### PR TITLE
3254 cleanup messages

### LINKED ISSUE #3254: Lua Documentation issues regarding messages
I find it very hard to understand and have trouble to distinguish between messages for message boxes and messages which arrive in the inbox. Maybe i do not understand all that stuff, so please be patient with my words :-) 

If possible at least the inbox related things should be renamed, e.g.:

Player:send_message() -> Player:send_notice()
Player.messages -> Player.notices
messages.lua:send_message() -> messages.lua:send_notice()
wl.Game.Message -> wl.Game.Notice

If renaming is too much work, it should be considered to clarify the descriptions of the functions, classes and property in that way that it is clear that those are only related to the inbox.

For message boxes there are 4 functions: 1 in wl.Game.Player and 3 in messages.lua. Those functions have some failures in the description:

https://wl.widelands.org/docs/wl/autogen_wl_game/#wl.game.Player.message_box
Here is a link to messages.lua:send_message() which is inbox related. So i guess this link is wrong? I do not understand why this function should be used carefully.

https://wl.widelands.org/docs/wl/autogen_auxiliary_messages/#message_box
The arguments 'message and 'parameters' are not explained much. By testing i found that at least this arguments must be passed in this way:

message_box(plr, "This is the title", p("This the message, or body"), {})

The message (or body) has to be in a p(), otherwise it isn't shown.
'parameters' must be at least an empty table, otherwise the game crashes.

https://wl.widelands.org/docs/wl/autogen_auxiliary_messages/#campaign_message_box
It says: "Sets message.h and message.w", i think this should be "Sets the width and height of the message box".
Message must be a table in form of {title = "The title", body = p("The body"),[opts]}. There should be an explanation of 'message' (which is not wl.Game.Message) somewhere so this can be linked.

https://wl.widelands.org/docs/wl/autogen_auxiliary_messages/#message_box_objective
It is unclear where one has to define obj_name, obj_title and obj_body.

Maybe we can omit some message box related functions?

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2018-01-25T12:02:16Z
 * owner: franku
 * the launchpad url was https://bugs.launchpad.net/bugs/1745359

### PR DESCRIPTION
**Pull request by frankystone**
Fixes #3254 

This should make distinguishing of inbox related functions and message box related function more clear.

I have also tried to make the documentation more clear and added an example for campaign_message_box. Where appropriate i linked to the corresponding functions  lua_game.cc <-> messages.lua.

### CODE REVIEW

File: data/scripting/messages.lua [hunk 1]
```diff
@@ -11,47 +11,48 @@ include "scripting/table.lua"
...
--- .. function:: send_message(player, title, body, parameters)
+-- .. function:: send_to_inbox(player, title, body, parameters)
 --
---    Sends a message to the player.
---    If the popup parameter is true and the player is in building mode,
---    the function waits until the player leaves the building mode
---    before sending the message (only in singleplayer)
+--    Sends a message to the inbox of a player.
+--    If the popup parameter is true and the player is in roadbuilding mode,
+--    the message is send after the player leaves the roadbuilding mode
```
Comment (gunchleoc) [inline data/scripting/messages.lua]:
```suggestion
--    the message is sent after the player leaves the roadbuilding mode
```

File: data/scripting/messages.lua [hunk 2]
```diff
@@ -82,20 +87,44 @@ function message_box(player, title, body, parameters)
...
---
---    :arg position: A string that indicates at which border of the screen the message box shall appear. Can be "top", "bottom", "right", "left" or a combination (e.g. "topright"). Overrides posx and posy. Default: Center. If only one direction is indicated, the other one stays central.
---    :arg scroll_back: If true, the view scrolls/jumps back to where it came from. If false, the new location stays on the screen when the message box is closed. Default: False.
---    :arg show_instantly: If true, the message box is shown immediately. If false, this function will wait until the player leaves the roadbuilding mode. Use this with care because it can be very interruptive. Default: false.
+--    Besides the normal message parameters (see :meth:`wl.game.Player.message_box`) the following ones can be used:
+--
+--    :arg position: A string that indicates at which border of the screen the message box shall appear.
+--                   Can be "top", "bottom", "right", "left" or a combination (e.g. "topright"). 
+--                   Overrides posx and posy. Default: Center. If only one direction is indicated, 
+--                   the other one stays central.
```
Comment (gunchleoc) [inline data/scripting/messages.lua]:
```suggestion
--                   the other one stays centered.
```

File: src/scripting/lua_game.cc [hunk 3]
```diff
@@ -300,17 +300,20 @@ int LuaPlayer::set_allow_additional_expedition_items(lua_State* L) {
...
+   .. method:: send_to_inbox(title, message[, opts])
 
-      Send a message to the player, the message will
-      appear in his inbox. Title or Message can be a
-      formatted using wideland's rich text.
+      Send a message to the players inbox.
+      There is also a function :meth:`send_to_inbox` in `messages.lua`, which 
+      has an option to wait before sending the message until the player leaves
+      the roadbuilding mode.
+      Title or Message can be a formatted using wideland's
```
Comment (gunchleoc) [inline src/scripting/lua_game.cc]:
```suggestion
      Title or Message can be a formatted using widelands' 
```

**Review by gunchleoc — COMMENTED** [review]
Some small nits

**Comment by gunchleoc** [comment]
Do you want to refactor this one as well?

`data/scripting/win_conditions/win_condition_functions.lua:      send_message(p, header, msg, options)`

You also need to update usages of `send_message` in the test suite.

There are also 2 codecheck failures:

```
src/scripting/lua_game.cc:306: Trailing whitespace at end of line
src/scripting/lua_game.cc:309: Trailing whitespace at end of line
```

**Comment by frankystone** [comment]
Does merging master change the files in the data-folder? I did always grepping over the data folder as well over src/scripting, and refactored the found files and committed the changes. But somehow the changes get reverted again from time to time and i have to refactor again :-( Maybe i do something wrong here...

I forgot the testsuite though, will refactor this soon.

Thanks for the quick review :)

File: data/scripting/messages.lua [hunk 4]
```diff
@@ -82,20 +87,44 @@ function message_box(player, title, body, parameters)
...
+--    :arg show_instantly: If true, the message box is shown immediately. If false, this function will
+--                    wait until the player leaves the roadbuilding mode. Use this with care because it
+--                    can be very interruptive. Default: false.
+--
+--    Example:
+-- .. code-block:: lua
+--
+--    campaign_message_box({title = "The title",    -- part of message
+--                          body = p("The body"),   -- part of message
+--                          200,                    -- optional sleeptime
```
Comment (Noordfrees) [inline data/scripting/messages.lua]: Are you sure that this will work?

Comment (frankystone) [reply data/scripting/messages.lua]: I have tested this with the debug console.

Comment (Noordfrees) [reply data/scripting/messages.lua]: `campaign_message_box({title="AAA",body="BBB",1000});print("XXX")` prints the message to the console directly after closing the box. I think there must be a key missing somewhere.

Comment (Noordfrees) [reply data/scripting/messages.lua]: Or `sleeptime` does not belong in the curly braces?

Comment (frankystone) [reply data/scripting/messages.lua]: Yes, you are right :-) I am fiddling with with REST markup right now to have a meaningful look in the documentation.

**Comment by gunchleoc** [comment]
This is how I do my merges (I do it on the console):

`git merge master`

Then either an editor comes up that I just close and I'm done, or I get some merge conflicts. If I get merge conflicts, I run another git command because it makes it easier to see the affected files:

`git status`

After fixing the code, I run `git status` again. It will show some files in red, let's say

`data/scripting/messages.lua`

I then run

`git add data/scripting/messages.lua`

or for adding the while directory, simply

`git add data`

Then I run

`git merge --continue`

**Comment by frankystone** [comment]
Thanks for your explanation. Don't know what i have done...

This is how it looks like now:
doc_campaign_message_box

File: data/scripting/messages.lua [hunk 5]
```diff
@@ -87,41 +87,48 @@ function message_box(player, title, body, parameters)
...
+--                          body = p("The body"),    -- text inside the window
+--                          w = 200,                 -- width (wl.game.Player.message_box())
+--                          h = 150,                 -- height (wl.game.Player.message_box())
 --                          position = "topleft",
---                          scroll_back = true
---                         })
+--                          field = scroll_to_field, -- see wl.game.Player.message_box()
+--                          scroll_back = true       -- only useful if 'field' was set
+--                         },
+--                          200                      -- optional sleeptime
```
Comment (gunchleoc) [inline data/scripting/messages.lua]:
```suggestion
--                         200                      -- optional sleeptime
```

**Review by gunchleoc — APPROVED** [review]
Found 1 small whitespace but in the documentation (see above).

Rest LGTM. I have tested the code snippet from the documentation and 1 win condition, starting condition,scenario, single player and multiplayer message.

**Comment by frankystone** [comment]
Thanks a lot for testing!

**Pull request merged**